### PR TITLE
CLDR-16253 Add component-specific parents to CLDR

### PR DESCRIPTION
**Pull request by sffc**
CLDR-16253

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

This is a proposed design for CLDR-16253. Spec text and ICU changes need to follow.

### CODE REVIEW

File: common/supplemental/supplementalData.xml [hunk 1]
```diff
@@ -5411,9 +5411,16 @@ XXX Code for transations where no currency is involved
 		<parentLocale parent="en_IN" locales="hi_Latn"/>
 		<parentLocale parent="es_419" locales="es_AR es_BO es_BR es_BZ es_CL es_CO es_CR es_CU es_DO es_EC es_GT es_HN es_MX es_NI es_PA es_PE es_PR es_PY es_SV es_US es_UY es_VE"/>
 		<parentLocale parent="fr_HT" locales="ht"/>
-		<parentLocale parent="no" locales="nb nn"/>
+		<parentLocale parent="no" locales="nb nn no_NO"/>
```
Comment (sffc) [inline common/supplemental/supplementalData.xml:5414]: This one has a comment in the ICU file:

```
            <!-- This one is a bit silly, it is just to generate a stub for no_NO, which is
                 not in CLDR. If we do not do this, then including it in localeIds will generate
                 empty no_Latn and no_Latn_NO and then no_NO aliasing to no_Latn_NO. -->
```

**Comment by sffc** [comment]
CC @hagbard

File: common/supplemental/supplementalData.xml (hunk 1 above)
Comment (macchiati) [reply common/supplemental/supplementalData.xml:5414]: Hmmm. let's bring that comment into the .xml file.

File: common/supplemental/supplementalData.xml [hunk 2]
```diff
@@ -5411,9 +5411,23 @@ XXX Code for transations where no currency is involved
...
 		<parentLocale parent="pt_PT" locales="pt_AO pt_CH pt_CV pt_FR pt_GQ pt_GW pt_LU pt_MO pt_MZ pt_ST pt_TL"/>
 		<parentLocale parent="zh_Hant_HK" locales="zh_Hant_MO"/>
+		<parentLocale alt="spoken" parent="ar_SA" locales="ars"/>
+	</parentLocales>
+
+	<parentLocales component="brkitr">
+		<parentLocale parent="zh" locales="zh_Hant"/>
+	</parentLocales>
+
+	<parentLocales component="coll">
```
Comment (macchiati) [inline common/supplemental/supplementalData.xml:5424]: Hmmm. There seems to be a general principle here. For collation, probably also for break iteration, we can do a fallback from a non-default script to the language-only locale. No objection to doing what you have here, but we might want to add a general structure later.

**Review by macchiati — DISMISSED** [review]
Only real request is to add the comment.

**Comment by sffc** [comment]
Any idea how to solve the CI maven build failure? My initial guess is that it is confused to see multiple `<parentLocales>` blocks?

**Comment by sffc** [comment]
Also see https://github.com/unicode-org/icu/pull/2294

**Comment by macchiati** [comment]
You are adding an attribute, so that needs to go into the dtd here 
https://github.com/unicode-org/cldr/blob/sffc-patch-3/common/dtd/ldmlSupplemental.dtd#L903

It will look something like:
<!ATTLIST parentLocales component NMTOKEN #IMPLIED >
    

However, it would be far cleaner — now that I look at it — for the component attribute values to match the top element involved. So you should change the above and your usage as follows.

coll ==> collations 
// as in https://github.com/unicode-org/cldr/blob/main/common/collation/am.xml#L14

brkitr ==> segmentations 
// as in https://github.com/unicode-org/cldr/blob/main/common/segments/ja.xml

**Review by macchiati — CHANGES_REQUESTED** [review]
However, it would be far cleaner — now that I look at it — for the component attribute values to match the top element involved. So you should change the above and your usage as follows.

coll ==> collations
// as in https://github.com/unicode-org/cldr/blob/main/common/collation/am.xml#L14

brkitr ==> segmentations
// as in https://github.com/unicode-org/cldr/blob/main/common/segments/ja.xml

**Comment by sffc** [comment]
@macchiati I changed PathHeader.txt as follows

```diff
-//supplementalData/parentLocales/parentLocale[@parent="%A"]/_locales			; Supplemental ; Locale ; Parent ; $1 ; HIDE
+//supplementalData/parentLocales/parentLocale[@parent="%A"][@alt="%A"]/_locales			; Supplemental ; Locale ; Parent ; $1 ; HIDE
```

But I'm still getting the following PathHeader errors

```
TestCompletenessNonLdmlDtd
Error:  (TestPathHeader.java:1313)  Error: //supplementalData/parentLocales/parentLocale[@parent="%A"]/_locales: Unknown path header[], Supplemental	Unknown	parentLocales	parentLocale[@parent="root"]/_locales
	No match in PathHeader.txt for //supplementalData/parentLocales/parentLocale[@parent="root"]/_locales
	You get only one message for all paths matching //supplementalData/parentLocales/parentLocale[@parent="%A"]/_locales
	For example, check to see if the field in PathHeader.txt is in PathHeader.PageId.
	If not, either correct PathHeader.txt or add it to PageId
	If you have a value attribute, you will need extra _ characters. The value attribute will show at the end with prefixed _, eg [...]/_skeleton.If there can be a value for the path then that element will add _. 

Error:  (TestPathHeader.java:1313)  Error: //supplementalData/parentLocales/parentLocale[@parent="%A"]/_alt: Unknown path header[], Supplemental	Unknown	parentLocales	parentLocale[@parent="ar_SA"]/_alt
	No match in PathHeader.txt for //supplementalData/parentLocales/parentLocale[@parent="ar_SA"]/_alt
	You get only one message for all paths matching //supplementalData/parentLocales/parentLocale[@parent="%A"]/_alt
	For example, check to see if the field in PathHeader.txt is in PathHeader.PageId.
	If not, either correct PathHeader.txt or add it to PageId
	If you have a value attribute, you will need extra _ characters. The value attribute will show at the end with prefixed _, eg [...]/_skeleton.If there can be a value for the path then that element will add _. 

Error:  (TestPathHeader.java:1313)  Error: //supplementalData/parentLocales[@component="%A"]/parentLocale[@parent="%A"]/_locales: Unknown path header[], Supplemental	Unknown	parentLocales[@component="segmentations"]	parentLocale[@parent="zh"]/_locales
	No match in PathHeader.txt for //supplementalData/parentLocales[@component="segmentations"]/parentLocale[@parent="zh"]/_locales
	You get only one message for all paths matching //supplementalData/parentLocales[@component="%A"]/parentLocale[@parent="%A"]/_locales
	For example, check to see if the field in PathHeader.txt is in PathHeader.PageId.
	If not, either correct PathHeader.txt or add it to PageId
	If you have a value attribute, you will need extra _ characters. The value attribute will show at the end with prefixed _, eg [...]/_skeleton.If there can be a value for the path then that element will add _. 
For help with DTD updates: https://cldr.unicode.org/development/updating-dtds
```

What I don't understand is that the error is giving me the old PathHeader.txt value of `//supplementalData/parentLocales/parentLocale[@parent="%A"]/_locales` even though I edited the txt file to add `[@alt="%A"]`. 🤔

**Comment by macchiati** [comment]
I'll look at this tomorrow

**Comment by macchiati** [comment]
It looks like you removed the line without the alt. Because the alt is optional, you need to have both lines.
Also, you need to have them in the right order, since the first match triggers. 

```
 - //supplementalData/parentLocales/parentLocale[@parent="%A"]/_locales			; Supplemental ; Locale ; Parent ; $1 ; HIDE
 + //supplementalData/parentLocales/parentLocale[@parent="%A"][@alt="%A"]/_locales			; Supplemental ; Locale ; Parent ; $1 ; HIDE

```

There is more about updating the DTDs in https://cldr.unicode.org/development/updating-dtds#h.thxta28vwqei.  If you see ways that can be improved, let me know and I can fix the problems!

**Review by macchiati — APPROVED** [review]

**Comment by macchiati** [comment]
Test failures: need to fix the following. Both are like the other test failures you fixed, but if you need help let me know.

```
 (TestDtdData.java:313)  Error: : isDistinguished supplementalData: elementName.equals("parentLocales") && attribute.equals("component"): expected false, got true
 (TestDtdData.java:313)  Error: : isDistinguished supplementalData: elementName.equals("parentLocale") && attribute.equals("alt"): expected true, got false
 (TestPathHeader.java:1313)  Error: //supplementalData/parentLocales/parentLocale[@parent="%A"]/_alt: Unknown path header[], Supplemental	Unknown	parentLocales	parentLocale[@parent="ar_SA"]/_alt
 Check failure on line 1313 in tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestPathHeader.java

(TestPathHeader.java:1313)  Error: //supplementalData/parentLocales[@component="%A"]/parentLocale[@parent="%A"]/_locales: Unknown path header[], Supplemental	Unknown	parentLocales[@component="segmentations"]	parentLocale[@parent="zh"]/_locales
 (TestPathHeader.java:1313)  Error: //supplementalData/parentLocales/parentLocale[@parent="%A"]/_alt: Unknown path header[], Supplemental	Unknown	parentLocales	parentLocale[@parent="ar_SA"]/_alt
 (TestPathHeader.java:1313)  Error: //supplementalData/parentLocales[@component="%A"]/parentLocale[@parent="%A"]/_locales: Unknown path header[], Supplemental	Unknown	parentLocales[@component="segmentations"]	parentLocale[@parent="zh"]/_locales
```

**Comment by sffc** [comment]
OK, the tests should be green now. But I needed to make the following changes:

1. I made `parentLocale`'s `alt` attribute be distinguishing. This seems to move it from being recognized as `_alt` to `[@alt="%A"]` in the PathHeader. The element already has a "value" field, which is `_locales`. It seems that if I wanted to add another "value" field, I would need to use a `_%E` path. However, I looked at other examples of "alt" attributes, and they were mostly all distinguishing.
2. I had to modify the `isDistinguishingOld` list in order to get that test to pass. I hope that's okay? Not sure how else to get the test to pass.

**Review by macchiati — CHANGES_REQUESTED** [review]

File: common/supplemental/supplementalData.xml [hunk 3]
```diff
@@ -5412,9 +5412,23 @@ XXX Code for transations where no currency is involved
 		<parentLocale parent="en_IN" locales="hi_Latn"/>
 		<parentLocale parent="es_419" locales="es_AR es_BO es_BR es_BZ es_CL es_CO es_CR es_CU es_DO es_EC es_GT es_HN es_MX es_NI es_PA es_PE es_PR es_PY es_SV es_US es_UY es_VE"/>
 		<parentLocale parent="fr_HT" locales="ht"/>
-		<parentLocale parent="no" locales="nb nn"/>
+		<parentLocale parent="no" locales="nb nn no_NO"/>
 		<parentLocale parent="pt_PT" locales="pt_AO pt_CH pt_CV pt_FR pt_GQ pt_GW pt_LU pt_MO pt_MZ pt_ST pt_TL"/>
 		<parentLocale parent="zh_Hant_HK" locales="zh_Hant_MO"/>
+		<parentLocale parent="ar_SA" locales="ars" alt="spoken"/>
```
[the remark this answers is not in the document]
Comment (macchiati) [reply common/supplemental/supplementalData.xml:5418]: @pedberg-icu

[the remark this answers is not in the document]
Comment (sffc) [reply common/supplemental/supplementalData.xml]: Here's the full comment from the ICU side

https://github.com/unicode-org/icu/blob/39dfee39b88a09fcafb33992a5495c4117490644/tools/cldr/cldr-to-icu/build-icu-data.xml#L365

```
            <!-- Some spoken languages (e.g. "ars") inherit all their data from a written language
                 (e.g. "ar_SA"). However CLDR doesn't currently support a way to represent that
                 relationship. Unlike deprecated languages for which an alias can be inferred from
                 the "languageAlias" CLDR data, there's no way in CLDR to represent the fact that
                 we want "ars" (a non-deprecated language) to inherit the data of "ar_SA".
                 This alias is the first example of potentially many cases where ICU needs to
                 generate an alias in order to affect "sideways inheritance" for spoken languages,
                 and at some stage it should probably be supported properly in the CLDR data. -->
            <forcedAlias source="ars" target="ar_SA"/>
```

[the remark this answers is not in the document]
Comment (sffc) [reply common/supplemental/supplementalData.xml]: For the purposes of this PR though I removed this so we can discuss it further. We may want to introduce more relationships like this, but maybe when we do, we should try to cover a broader set of spoken languages.

**Review by macchiati — DISMISSED** [review]

**Comment by macchiati** [comment]
I think the ticket just needs squashing to fix the commit message, and then can be merged.

**Review by macchiati — APPROVED** [review]

**Pull request merged**